tests/AppTest.php

```diff
@@ -67,4 +68,123 @@ public function testUnexpectedOutputLateError(): void
             ob_end_clean();
         }
     }
+
+    public function provideUrlBuildingCases(): iterable
+    {
+        return [
+            // simple cases
+            [[], false, [], '/', '/index.php', '/default.php8', '/'],
+            [[], false, [], '/test/', '/test/index.php', '/test/default.php8', '/test/'],
+            [[], true, [], '/', '/index.php', '/default.php8', '/'],
+            [[], true, [], '/test/', '/test/index.php', '/test/default.php8', '/test/'],
+
+            // simple cases with query args in request
+            // [[], false, [], '/?test=atk4', '/index.php?test=atk4', '/default.php8?test=atk4', '/?test=atk4'],
+            // [[], false, [], '/test/?test=atk4', '/test/index.php?test=atk4', '/test/default.php8?test=atk4', '/test/?test=atk4'],
+            [[], true, [], '/?test=atk4', '/index.php?test=atk4', '/default.php8?test=atk4', '/?test=atk4'],
+            [[], true, [], '/test/?test=atk4', '/test/index.php?test=atk4', '/test/default.php8?test=atk4', '/test/?test=atk4'],
+
+            // simple cases with extra query args in request
+            // [[], false, ['extra_args' => 'atk4'], '/?test=atk4', '/index.php?extra_args=atk4&test=atk4', '/default.php8?extra_args=atk4&test=atk4', '/?extra_args=atk4&test=atk4'],
+            // [[], false, ['extra_args' => 'atk4'], '/test/?test=atk4', '/test/index.php?extra_args=atk4&test=atk4', '/test/default.php8?extra_args=atk4&test=atk4', '/test/?extra_args=atk4&test=atk4'],
+            [[], true, ['extra_args' => 'atk4'], '/?test=atk4', '/index.php?extra_args=atk4&test=atk4', '/default.php8?extra_args=atk4&test=atk4', '/?extra_args=atk4&test=atk4'],
+            [[], true, ['extra_args' => 'atk4'], '/test/?test=atk4', '/test/index.php?extra_args=atk4&test=atk4', '/test/default.php8?extra_args=atk4&test=atk4', '/test/?extra_args=atk4&test=atk4'],
+
+            // simple cases with page as string
+            ['test', false, [], '/', 'test.php', 'test.php8', 'test'],
+            ['test/test/a', false, [], '/', 'test/test/a.php', 'test/test/a.php8', 'test/test/a'],
+            ['test/index', false, [], '/', 'test/index.php', 'test/index.php8', 'test/index'],
+            ['test/index', false, [], '/test/', 'test/index.php', 'test/index.php8', 'test/index'],
+            ['test', true, [], '/', '/index.php', '/default.php8', '/'],
+            ['test', true, [], '/request-url', '/request-url.php', '/request-url.php8', '/request-url'],
```

Comment:
brokes testing on Windows

Reply:
I will check later on Windows the directory separator is probably messed up, but I think only in tests due to "emulation" of the webserver

Reply:
to be fixed in https://github.com/atk4/ui/pull/2095#discussion_r1326492320 thread